cache/models/samples.js

```diff
@@ -1250,153 +1269,263 @@ module.exports = {
+              keys = modelUtils.applyLimitAndOffset(opts, filteredKeys);
+            } else if (opts.order === ['-name']) {
+              filteredKeys.sort();
+              filteredKeys.reverse();
+              keys = modelUtils.applyLimitAndOffset(opts, filteredKeys);
+            }
+            debug('has name filter only. sorted and applied' +
+              ' limit/offset. Samples arr size: %d', keys.length);
+          }
+
+          return getSamplesAndAspects(keys);
+        })
+        .then((redisResponses) => {
+          const { samples, sampAspectMap } = assembleSampleAspects(
+            redisResponses);
+          const fSamples = modelUtils.applyFiltersOnSampleObjs(samples,
+            opts);
+          return applyAttributesFilter(fSamples, opts,
             sampAspectMap, req.method);
         });
-    }
+    } // featureToggle end
 
-    /**
-     * Name filter and no wildcards:
-     * If {filter: { name: 's1|a1'}}
-     *  get one sample
-     *
-     * If { filter: { name: 's1|a1, s2|a2'}}
-     *  get the samples with comma separated names
-     *  If default name order, sort the array of sample names
-     *  Apply limit and offset
-     *  Get samples
-     *
-     * Then,
-     * If order other than name, then sort and apply limit/offset
-     * apply attributes filter, cleanup and return
-     *
+    /*
+     * Send a batch of redis commands to get all the samples sorted
+     * lexicographically by key (i.e. sample name). If there are no filters,
+     * pass the limit and offset through as part of the initial redis command.
+     * If there are filters, we need to load more records so we can return the
+     * right number of records in case some get filtered out later.
      */
-    if (hasNameFilterOnly && !opts.filter.name.includes('*')) {
-      let nameFilterArr = opts.filter.name.split(',').map((item) =>
-        item.trim());
-      return Promise.resolve()
-        .then(() => {
+    const sortArgs = [constants.indexKey.sample, 'alpha'];
+    const hasFilters = Object.keys(opts.filter).length > 0;
+    if (!hasFilters) {
+      sortArgs.push('LIMIT', opts.offset, opts.limit);
+    }
+
+    return Promise.resolve()
+      .then(() => {
+        // If there is a name param with no wildcards, get the sample directly.
+        const nameFilter = opts.filter.name;
+        if (nameFilter && !nameFilter.includes('*')) {
+          const nameFilterArr = opts.filter.name.split(',').map((item) =>
+            item.trim());
           if (nameFilterArr.length === 1) { // one sample
             return getOneSample(nameFilterArr[0])
               .then(([samp, asp]) => samp && asp ? [samp, asp] : []);
           }
 
           if (!hasOrder || opts.order === ['name']) {
             nameFilterArr.sort();
-            nameFilterArr = modelUtils.applyLimitAndOffset(opts, nameFilterArr);
           }
 
           // multiple samples
           const sampleKeys = nameFilterArr.map((name) =>
```

Comment:
I introduce the option to get samples by comma separated name (even without toggle enabled). Helps with passing all the common tests (which include comma separated names) and it will help with users like Kyle who are only requesting specific number of samples for their integration.